parser/src/parse.rs

```diff
@@ -0,0 +1,434 @@
+                        }
+                        '\'' => {
+                            if in_double_quotes {
+                                res.push(c);
+                            }
+                        }
+                        _ => {
+                            res.push(c);
+                        }
+                    };
+                    res.push(c2);
+                    continue;
+                }
+                _ => {
+                    res.push(c);
+                    continue;
+                }
+            };
+        } else if c == '\'' {
+            if in_double_quotes {
+                res.push(c);
+                continue;
+            }
+
+            in_single_quotes = !in_single_quotes;
+            continue;
+        } else if c == '"' {
+            if in_single_quotes {
+                res.push(c);
+                continue;
+            }
+
+            in_double_quotes = !in_double_quotes;
+            continue;
+        } else if !in_quotes {
+            return Err(ParseError::MissingOpeningQuote);
+        }
+
+        res.push(c);
+    }
+
+    Ok(res)
+}
+
+fn parse_quoted_string(
+    s: &str,
+    mut chars: &mut Enumerate<Chars>,
+    mut res: String,
+    quote: char,
+) -> Result<String, ParseError> {
+    let mut in_single_quotes = quote == '\'';
+    let mut in_double_quotes = quote == '"';
+
+    while let Some((idx, c)) = chars.next() {
+        let in_quotes = in_single_quotes || in_double_quotes;
+
+        if c == '\\' && in_quotes {
+            match chars.next() {
+                None => {
+                    return Err(ParseError::InvalidEscape {
+                        escape: format!("{}", c),
+                        index: idx,
+                        string: String::from(s),
+                    });
+                }
+                Some((idx, c2)) => {
+                    let mut push_escape_character = false;
+
+                    let value = match c2 {
+                        'a' => '\u{07}',
+                        'b' => '\u{08}',
+                        'v' => '\u{0B}',
+                        'f' => '\u{0C}',
+                        'n' => '\n',
+                        'r' => '\r',
+                        't' => '\t',
+                        '\\' => c2,
+                        '?' => c2,
+                        '\'' => {
+                            push_escape_character = in_double_quotes;
+                            c2
+                        }
+                        '"' => {
+                            push_escape_character = in_single_quotes;
+                            c2
+                        }
+                        '`' => c2,
+                        'x' | 'u' | 'U' => {
+                            let length = match c2 {
+                                'x' => 2,
+                                'u' => 4,
+                                'U' => 8,
+                                _ => unreachable!(),
```

Comment:
Not sure if there's a better way to ensure that this second match is exhaustive - after all we know for sure that this should not happen.

Reply:
I think this is fine. I messed with it for a few minutes and concluded it was just a limitation of the type system and what we were doing. As long as `c2` is being used in both match arms, then we can confidently say that that branch really is unreachable.